src/plan_expand_hypertable.c

```diff
@@ -70,6 +72,124 @@ is_chunk_exclusion_func(Expr *node)
 	return false;
 }
 
+static bool
+is_time_bucket_function(Expr *node)
+{
+	if (IsA(node, FuncExpr) &&
+		strncmp(get_func_name(castNode(FuncExpr, node)->funcid), "time_bucket", NAMEDATALEN) == 0)
+		return true;
+
+	return false;
+}
+
+/*
+ * Transform time_bucket calls of the following form in WHERE clause:
+ *
+ * time_bucket(width, column) > value
+ * column > value - width
+ *
+ * time_bucket(width, column) < value
+ * column < value + width
+ *
+ * This type of WHERE clause is used when querying continuous aggregation
+ * views.
+ */
+static OpExpr *
+transform_time_bucket_comparison(PlannerInfo *root, OpExpr *op)
+{
+	Expr *left = linitial(op->args);
+	Expr *right = lsecond(op->args);
+
+	FuncExpr *time_bucket = castNode(FuncExpr, (IsA(left, FuncExpr) ? left : right));
+	Expr *value = IsA(right, Const) ? right : left;
+
+	Const *width = linitial(time_bucket->args);
+	Oid opno = op->opno;
+	Oid subst_opno;
+	TypeCacheEntry *tce;
+	int strategy;
+	char *opname;
+
+	/*
+	 * if time_bucket call is on wrong side we switch operator
+	 */
+	if (IsA(right, FuncExpr))
+		opno = get_commutator(op->opno);
+
+	if (!OidIsValid(opno))
+		return op;
+
+	tce = lookup_type_cache(exprType((Node *) time_bucket), TYPECACHE_BTREE_OPFAMILY);
+	strategy = get_op_opfamily_strategy(opno, tce->btree_opf);
+
+	if (strategy == BTLessStrategyNumber || strategy == BTLessEqualStrategyNumber)
+		opname = "+";
+	else if (strategy == BTGreaterStrategyNumber || strategy == BTGreaterEqualStrategyNumber)
+		opname = "-";
+	else
+		return op;
+
+	subst_opno = get_operator(opname,
+							  PG_CATALOG_NAMESPACE,
+							  exprType((Node *) value),
+							  exprType((Node *) width));
+
+	if (OidIsValid(subst_opno))
+	{
+		Oid resulttype = get_op_rettype(subst_opno);
+		Expr *subst = make_opclause(subst_opno,
+									tce->type_id,
+									false,
+									value,
+									(Expr *) width,
+									InvalidOid,
+									InvalidOid);
+
+		/*
+		 * check if resulttype of operation returns correct datatype
+		 *
+		 * date OP interval returns timestamp so we need to insert
+		 * a cast to keep toplevel expr intact when datatypes don't match
+		 */
+		if (tce->type_id != resulttype)
+		{
+			Oid cast_func = get_cast_func(resulttype, tce->type_id);
+
+			if (!OidIsValid(cast_func))
+				return op;
+
+			subst = (Expr *)
+				makeFuncExpr(cast_func, tce->type_id, list_make1(subst), InvalidOid, InvalidOid, 0);
+		}
+		else if (tce->type_id == TIMESTAMPTZOID && width->consttype == INTERVALOID)
+		{
+			/*
+			 * TIMESTAMPTZ OP INTERVAL is marked stable and unsafe
+			 * to evaluate at plan time unless it only has a time
+			 * component
+			 */
+			Interval *interval = DatumGetIntervalP(width->constvalue);
+
+			if (interval->day == 0 && interval->month == 0)
+				subst = (Expr *) estimate_expression_value(root, (Node *) subst);
```

Comment:
don't we need to return op in an else? won't this just chug merrily along otherwise?

Reply:
thats intentional because estimate_expression_value is not safe for those but subst still has  a valid expression we can use we just dont try to estimate it